js/private/js_run_devserver.js

```diff
@@ -0,0 +1,212 @@
+                })
+            )
+            return totalSynced
+        } else {
+            if (process.env.JS_BINARY__LOG_DEBUG) {
+                console.error(
+                    `Syncing file ${src.slice(RUNFILES_ROOT.length + 1)}`
+                )
+            }
+            if (exists) {
+                await fs.promises.unlink(dst)
+            } else {
+                mkdirpSync(path.dirname(dst))
+            }
+            await fs.promises.copyFile(src, dst)
+            return 1
+        }
+    } catch (e) {
+        console.error(e)
+        process.exit(1)
+    }
+}
+
+// Sync list of files to the sandbox
+async function sync(files, sandbox) {
+    let totalSynced = 0
+    console.error('Syncing...')
+    const startTime = perf_hooks.performance.now()
+    await Promise.all(
+        files.map(async (file) => {
+            const src = path.join(RUNFILES_ROOT, file)
+            const dst = path.join(sandbox, file)
+            const numSynced = await syncRecursive(src, dst)
+            totalSynced += numSynced
+        })
+    )
+    var endTime = perf_hooks.performance.now()
+    console.error(
+        `${totalSynced} file${
+            totalSynced > 1 ? 's' : ''
+        } synced in ${Math.round(endTime - startTime)} ms`
+    )
+}
+
+async function main(args, sandbox) {
+    console.error(
+        `\n\nStarting js_run_devserver ${process.env.JS_BINARY__TARGET}`
+    )
+
+    const configPath = path.join(RUNFILES_ROOT, args[0])
+
+    const config = JSON.parse(await fs.promises.readFile(configPath))
+
+    await sync(config.data_files, sandbox)
+
+    return new Promise((resolve) => {
+        const cwd = process.env.JS_BINARY__CHDIR
+            ? path.join(sandbox, process.env.JS_BINARY__CHDIR)
+            : sandbox
+
+        const tool = config.tool
+            ? path.join(RUNFILES_ROOT, config.tool)
+            : config.command
+
+        const toolArgs = args.slice(1)
+
+        console.error(`Running '${tool} ${toolArgs.join(' ')}' in ${cwd}\n\n`)
+
+        const proc = child_process.spawn(tool, toolArgs, {
+            cwd: cwd,
+            stdio: 'inherit',
+            env: {
+                ...process.env,
+                BAZEL_BINDIR: '.',
+                JS_BINARY__CHDIR: '',
+            },
+        })
+
+        proc.on('close', (code) => {
+            console.error(`child tool process exited with code ${code}`)
+            resolve()
+            process.exit(code)
+        })
+
+        process.stdin.on('data', async (chunk) => {
+            try {
+                const chunkString = chunk.toString()
+                if (chunkString.includes('IBAZEL_BUILD_COMPLETED SUCCESS')) {
+                    if (process.env.JS_BINARY__LOG_DEBUG) {
+                        console.error('IBAZEL_BUILD_COMPLETED SUCCESS')
+                    }
+                    // Re-parse the config file to get the latest list of data files to copy
+                    await sync(
```

Comment:
This can get triggered multiple times and then have multiple overlapping `sync`s. Is that ok? Since we are only creating+updating files it might be but if we are deleting files it won't be.

This is where if we want to ensure there are no overlapping `sync`s we'd do something such as:

```
let synching = Promise.resolve();
...
await (synching = synching.then(()( => sync()))
```

Reply:
🧙